From eff67fed8487f4f4f201166f415e286b885e3326 Mon Sep 17 00:00:00 2001
From: Stefan Lippers-Hollmann <s.l-h@gmx.de>
Date: Wed, 28 Jan 2015 05:25:42 +0100
Subject: [PATCH] x86: add ASRock Q1900DC-ITX mainboard reboot quirk

The ASRock Q1900DC-ITX mainboard (Baytrail-D) hangs randomly in both
BIOS and UEFI mode while rebooting unless reboot=pci is used. Add a
quirk to reboot via the pci method.

Signed-off-by: Stefan Lippers-Hollmann <s.l-h@gmx.de>
Cc: stable@kernel.org
---

The problem is very intermittent and hard to debug, it might succeed
rebooting just fine 40 times in a row - but fails half a dozen times
the next day. It seems to be slightly less common in BIOS CSM mode
than native UEFI (with the CSM disabled), but it does happen in either
mode.

Most of the time it already hangs during POST, but occassionally it
might even make it through the bootloader and the kernel might even
start booting, but then hangs before the mode switch. The same symptoms
occur with grub-efi, gummiboot and grub-pc, just as well as (at least)
kernel 3.16-3.19 (I haven't tried older kernels than 3.16). Upgrading
to the most current mainboard firmware of the ASRock Q1900DC-ITX,
version 1.20, does not improve the situation.

Searching the web seems to suggest that other Bay Trail-D mainboards
might be affected as well.

 arch/x86/kernel/reboot.c | 10 ++++++++++
 1 file changed, 10 insertions(+)

--- a/arch/x86/kernel/reboot.c
+++ b/arch/x86/kernel/reboot.c
@@ -183,6 +183,16 @@ static struct dmi_system_id __initdata r
 		},
 	},
 
+	/* ASRock */
+	{	/* Handle problems with rebooting on ASRock Q1900DC-ITX */
+		.callback = set_pci_reboot,
+		.ident = "ASRock Q1900DC-ITX",
+		.matches = {
+			DMI_MATCH(DMI_BOARD_VENDOR, "ASRock"),
+			DMI_MATCH(DMI_BOARD_NAME, "Q1900DC-ITX"),
+		},
+	},
+
 	/* ASUS */
 	{	/* Handle problems with rebooting on ASUS P4S800 */
 		.callback = set_bios_reboot,
